draft-ion.adoc

```diff
@@ -208,79 +200,176 @@ represented as follows:
+
+If a member name is not equal to the case-sensitive octet sequence `meta`, an Ion parser MUST NOT identify a member as a _Meta Object_.
+
+The member names within an identified _Meta Object_ MUST be unique; Ion parsers MUST either
+reject _Meta Object_ members with duplicate member names or use a JSON
+parser that returns only the lexically last duplicate member, as specified
 in {uri-ecmaScript-5-1-15-12}[Section 15.12 (The JSON Object)] of {uri-ecmaScript-5-1}[ECMAScript 5.1].
 
-The set of fields that a _Meta Object_ must contain to be considered valid
+The set of members that a _Meta Object_ must contain to be considered valid
 is resource-dependent.  In the absense of any resource-specific requirements,
-any field not understood by implementations MUST be ignored.
+any member not understood by implementations MUST be ignored.
+
+[[links]]
+== Links
+
+An Ion Link is an Ion Object where its `meta` member contains an {uri-IRIs}[IRI] member named `href`. An Ion Link is therefore JSON serialization of a {uri-rfc5988}[web link] to another resource.
+
+Ion parsers MUST identify any JSON structure matching this definition as an Ion Link.  Anything that does not match this defined structure MUST NOT be considered an Ion Link.
+
+A non-normative example of a minimal Ion link:
+
+.Example simplest Ion link:
+[source,json]
+----
+{ "meta": { "href": "https://api.example.com/corporations/acme" } }
+----
+
+=== Link Relation Types
+
+All Ion links have one or more <<links, link relation types>>.  A link's total set of link relation types is the set union of the link's implicit link relation type and the link's explicit link relation types.
+
+==== Implicit Link Relation Type
+
+All Ion links have exactly one implicit link relation type.  The implicit link relation type can be one of three possible values depending on where the link is located in the Ion content:
+
+1. Root Object
+2. Named Link
+3. Collection Resource `items` array element
+
+===== Root Object Implicit Link Relation Type
+
+If the Root Object is also found to be an Ion Link, it has an implicit link relation type of `self`.
+
+A non-normative example of a Root Object that is also an Ion Link:
+
+[source,json]
+----
+{
+  "meta": { "href": "https://api.example.com/users/1" },
+  "name": "Joe"
+}
+----
 
-=== Registered Metadata Field Names
+===== Named Link Implicit Link Relation Type
 
-The following field names are registered in the IANA Ion Metadata field registry
-defined in [TBD].  None of the fields defined below are intended to be mandatory
+If a JSON Object member value is an Ion Link, the Ion Link is implicitly assigned a link relation type that equals the member name.
+
+A non-normative example:
+
+[source,json]
+----
+{
+  "employer": { "meta": { "href": "https://api.example.com/corporations/acme" } }
+}
+----
+
+The above example shows an Ion Link member named `employer`.  The octet sequence `employer` is therefore assigned as the link's implicit link relation type.
```

Comment:
Great question. So, the answer is how do I get evolvability: by what is in the message or how I consume the message. There is a "Hypermedia Contract" or "Promise". That promise is that if you get this thing, this is what it will mean and if I change what it means, I am breaking the contract. These issues and your example are completely outside the realm of the media-type IMO in one sense. In another, Ion as a media-type can facilitate the change using some deprecation hints and the ability to send new and old semantics in the same message, etc., which are things to consider adding to address what I think you are grappling with.

Further, there are 2 types of semantics involved: semantics of the media-type which if you change, you have broken the contract of the spec and semantics of the domain, which if you change or remove, you break the contract of the domain.

Your examples are the latter (domain semantics) and actually subject to different constraints than the media-type contract. So, trying to make sure people don't do things that break domain things in the spec is mandating how they should implement a good evolvable API. I think the lines are blurring here and there is an aspect of education on hypermedia design principles feeling like they are being baked into the spec. But I don't think they belong there.

So, in your example, if I rely on some semantic element as a client, I don't just assume it is there in the code and have a failure path (which may be log alerts like mad cause something I am depending on did not show up and gracefully fail). The media-type could contain metadata to relate something from before to something new or facilitate transition, use profiles like ALPS or your API could use profiles and other things to give machines information. But it is not the media-type's job to enforce the domain contract.

Hope this is making sense. The Ion Document should have not dramatic structural changes from the meaning of Ion Elements, but a message can always have dramatic structural changes from a domain standpoint (to the pain of all parties involved if not done wisely).

This is like a browser expecting some domain element to be there and blowing up if it is not that is supra-HTML. You could do that, but the browser (client) is no longer implementing just the media-type, but it is implementing the domain.

You as a human may be looking for some domain element as the ultimate client (like a program being the ultimate client that uses a hypermedia client indirectly to process a message). If something is not there you may not be happy, but you do something else to recover.

This spec is just about what a code based client can expect in the message semantics with respect to the message itself. It knows nothing about the domain or what you should or should not do with the domain specific elements other than say this is meta data, this is a control, this is data. If some domain semantic element is not there a "ion client" does not care because it programmed to the spec. But the ultimate client using that may very well care and so interacts with the message assuming nothing.

```
if what I want is here
  # do this
else
  # gracefully recover
end
```

That is how I think about this. No silver bullet for people doing bad things with APIs. But if you consume these messages correctly and providers evolve  their API correctly, you get the desired hypermedia benefit. 

Make any sense?